core/controllers/domain_objects_validator.py

```diff
@@ -243,3 +243,67 @@ def validate_suggestion_images(files):
     # The files argument do not represent any domain class, hence dict form
     # of the data is returned from here.
     return files
+
+
+def validate_exploration_change(exp_change_dict):
+    """Validate the exploration change list
+
+    Args:
+        exp_change_dict: dict. The exploration change dictionary
+            that needs to be validated.
+
+    Returns:
+        exp_domain.ExplorationChange. The exploration change domain
+        object.
+    """
+    if exp_change_dict.get('cmd') is None:
+        raise base.BaseHandler.InvalidInputException(
+            'Missing cmd key in change dict')
+
+    # Validate state properties.
+    if exp_change_dict['cmd'] == exp_domain.CMD_EDIT_STATE_PROPERTY:
+        if (
+            exp_change_dict['property_name'] ==
+            exp_domain.STATE_PROPERTY_INTERACTION_ANSWER_GROUPS
+        ):
+            answer_groups = exp_change_dict['new_value']
+            for answer_group in answer_groups:
+                if answer_group['tagged_skill_misconception_id'] is not None:
+                    raise base.BaseHandler.InvalidInputException(
+                        'tagged_skill_misconception_id of the answer '
+                        'group should be None.'
+                    )
+                if len(answer_group['rule_specs']) == 0:
+                    raise base.BaseHandler.InvalidInputException(
+                        'The answer group should contain atleast one rule spec.'
+                    )
+                if answer_group['outcome']['dest'] == '':
+                    raise base.BaseHandler.InvalidInputException(
+                        'The destination for the answer group is not valid.'
+                    )
+                if (
+                    answer_group['outcome']['labelled_as_correct'] is True and
+                    answer_group['outcome']['dest'] == exp_change_dict[
```

Comment:
why are we adding these checks for change dict. generally we add these validations as part of exp/state validation?

Reply:
Same reason as above, please let me know if I missed something here.

Reply:
This comment is not about the placement of this check, it's about why we need this check, don't we already have this check in the State class? In general, we don't have such data validations for change command, right? I mean we only validate the schema of change dict and we validate internal data after applying the change_dict to exploration using exploration.validate()

Is there any specific reason why we are adding this check? (I'm assuming we also have these checks in State class)

Reply:
Yes, we do have validations present inside the State class. As far as I know we are currently adding 2 layers of validation. One in the controller layer and another with the domain layer. We do validate the data inside the change command at the controller layer as well as at domain layer. This particular file calls the validation method for the data that is present in the change dict.
Thanks

Reply:
In the controller layer we only validate the basic structure/schema and all other data-related validations exist in the State class. Do we have [all these](https://github.com/oppia/oppia/blob/develop/core/domain/state_domain.py#L1779) validations in the controller layer?

In general, I think having these validations in multiple places creates duplication which will be hard to maintain in the future! Let me know what you think.

Reply:
I agree with you, so basically we do not directly have the validation for all of them in the controller layer but we call their validate method from the controller layer. As my changes were few so I directly checked them here.
Should I remove this complete part of validation?

Reply:
I think we sometimes have redundancy in some of the validations. The main thing for controller validation is that it is only possible to do the validations separate for each arg (in some cases that means we can only validate it partially as for the full validation we would need some other arg (like the version)).

Also, the validation in the controller layer is important to make sure that the args are correctly typed. I agree that some of the additional stuff is not needed here and we should do only the basic validation here and the other stuff elsewhere.

Reply:
> Please let me know what should be the appropriate next steps I should follow.

I think you can just leave the controller layer check as is and add additional checks in the relevant class validation method!

(Note: The change_dict does go through the basic check in the existing structure!)

Reply:
Done